crates/sui-proxy/src/prom_to_mimir.rs

```diff
@@ -0,0 +1,387 @@
+    type Item = mimir::LabelPair;
+    type IntoIter = std::vec::IntoIter<Self::Item>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.state.into_iter()
+    }
+}
+
+impl From<&Counter> for Mimir<mimir::Sample> {
+    fn from(c: &Counter) -> Self {
+        // timestamp_ms does not exist in prom::Metric, so we mark "now" as an approximation
+        // the actual time could be now - (interval + jitter)
+        let now = SystemTime::now().duration_since(UNIX_EPOCH).unwrap();
+        Self {
+            state: mimir::Sample {
+                value: c.get_value(),
+                timestamp_ms: now.as_millis() as i64,
+                ..Default::default()
+            },
+        }
+    }
+}
+impl From<&Gauge> for Mimir<mimir::Sample> {
+    fn from(c: &Gauge) -> Self {
+        // timestamp_ms does not exist in prom::Metric, so we mark "now" as an approximation
+        // the actual time could be now - (interval + jitter)
+        let now = SystemTime::now().duration_since(UNIX_EPOCH).unwrap();
+        Self {
+            state: mimir::Sample {
+                value: c.get_value(),
+                timestamp_ms: now.as_millis() as i64,
+                ..Default::default()
+            },
+        }
+    }
+}
+impl Mimir<mimir::Sample> {
+    fn sample(self) -> mimir::Sample {
+        self.state
+    }
+}
+
+impl From<&Histogram> for Mimir<mimir::Histogram> {
+    fn from(h: &Histogram) -> Self {
+        info!("{:?}", h); // we only log histograms for now, no implementation to convert them yet
+        Self {
+            state: mimir::Histogram::default(),
+        }
+    }
+}
+impl Mimir<mimir::Histogram> {
+    fn histogram(self) -> mimir::Histogram {
+        self.state
+    }
+}
+impl From<Vec<MetricFamily>> for Mimir<Vec<mimir::WriteRequest>> {
+    fn from(metric_families: Vec<MetricFamily>) -> Self {
+        let mut requests = vec![];
+        for metric in metric_families {
+            // TODO stop using state directly
+            let ts =
+                RepeatedField::<mimir::TimeSeries>::from_vec(Mimir::from(metric.clone()).state);
+
+            // TOOD add From impl
+            let mt = match metric.get_field_type() {
+                MetricType::COUNTER => mimir::MetricMetadata_MetricType::COUNTER,
+                MetricType::GAUGE => mimir::MetricMetadata_MetricType::GAUGE,
+                MetricType::HISTOGRAM => mimir::MetricMetadata_MetricType::HISTOGRAM,
+                MetricType::SUMMARY => mimir::MetricMetadata_MetricType::SUMMARY,
+                MetricType::UNTYPED => mimir::MetricMetadata_MetricType::UNKNOWN,
+            };
+
+            // filter out the types we don't support
+            match mt {
+                mimir::MetricMetadata_MetricType::COUNTER
+                | mimir::MetricMetadata_MetricType::GAUGE => (),
+                other => {
+                    warn!("{:?} is not yet implemented, skipping metric", other);
```

Comment:
this would be a lot of logs?  I'd change it to debug so we can toggle it to see what's happening when we need to.

Reply:
will do, it's just left in there from testing today.